### PR TITLE
Onboarding WPF's Grid control

### LINKED ISSUE #2138: Grid layout is broken in some more complex cases with * width
Grid layout is broken in some more complex cases with * width 
examples:

- Path Bounds go outside first column, although path should always fit into first grid column

```XAML
    <Grid ColumnDefinitions="*,*">
        <Path Data="M0,0 L0,10 10,10 10,0Z" Fill="Red" Stroke="Blue" StrokeThickness="5" Stretch="Uniform" Grid.Column="0" 
              HorizontalAlignment="Center" VerticalAlignment="Center" Margin="10" />
        <Border Background="Yellow" BorderBrush="Black" BorderThickness="2" Grid.Column="1" />
    </Grid>
```

- second grid column with static (100 pixels) width is not visible when parent width is decreased during resize

```XAML
    <Grid ColumnDefinitions="*,100" HorizontalAlignment="Left">
        <Path Data="M0,0 L0,10 10,10 10,0Z" Fill="Red" Stroke="Blue" StrokeThickness="5" Stretch="Uniform" Grid.Column="0" 
              HorizontalAlignment="Center" VerticalAlignment="Center" Margin="10" />
        <Border Background="Yellow" BorderBrush="Black" BorderThickness="2" Grid.Column="1" />
    </Grid>
```

related issue #2129

### PR DESCRIPTION
**Pull request by jmacato**
## What does the pull request do?

Ports the Grid control from dotnet/wpf repository. 
## What is the updated/expected behavior with this PR?

`Grid`'s behavior should now be inline with WPF's, hence fixing some long-running issues with our own `Grid` control. 

## Checklist

- [x] Added unit tests (if possible)? 

## Fixed issues
Fixes #2138 (i tested the described scenarios by @donandren but this needs another pair of eyes).

### CODE REVIEW

File: src/Avalonia.Controls/GridWPF.cs [hunk 1]
```diff
@@ -0,0 +1,3854 @@
...
+                ResolveStarLegacy(definitions, availableSize);
+            }
+            else
+            {
+                ResolveStarMaxDiscrepancy(definitions, availableSize);
+            }
+        }
+
+        // original implementation, used from 3.0 through 4.6.2
+        private void ResolveStarLegacy(
```
Comment (Gillibald) [inline src/Avalonia.Controls/GridWPF.cs]: We should probably leave out the old implementation

**Comment by jmacato** [comment]
WPF's Grid running on ControlCatalog (still broken though)
image

**Comment by jmacato** [comment]
Getting there! (calendar broke for some reason)
image

**Comment by Gillibald** [comment]
I don't think we should add class handlers for column, row attached to each visual. Instead do this when the property is attached to a visual.

**Comment by jmacato** [comment]
@Gillibald Yes, thats the right way. There's still some temporary hacks and that is one of 'em so thanks for the reminder :smile:

**Comment by jmacato** [comment]
`ShowGridLines` now working
image

**Comment by jmacato** [comment]
Almost all features but `SharedSizeScopes` is working now :smile:

File: src/Avalonia.Controls/Grid/GridLinesRenderer.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+{
+    internal class GridLinesRenderer : Control
+    {
+        /// <summary>
+        /// Static initialization
+        /// </summary>
+        static GridLinesRenderer()
+        {
+            var oddDashArray = new List<double>();
+            oddDashArray.Add(_dashLength);
```
Comment (Gillibald) [inline src/Avalonia.Controls/Grid/GridLinesRenderer.cs]: even and odd are the same

Comment (jmacato) [reply src/Avalonia.Controls/Grid/GridLinesRenderer.cs]: Nice catch! Double-checked the original source, i've mangled it while copypasting it seems

**Comment by jmacato** [comment]
Grid is now fully-functional :label: 
image

**Comment by x2bool** [comment]
Great job! I hope all of those small bugs of the old Grid implementation will be gone now.

File: src/Avalonia.Controls/Grid/DefinitionBase.cs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+        /// <summary>
+        /// Callback to notify about entering model tree.
+        /// </summary>
+        internal void OnEnterParentTree(Grid grid, int index)
+        {
+            Parent = grid;
+            _parentIndex = index;
+        }
+
+        internal void UpdateSharedScope()
```
Comment (jmacato) [inline src/Avalonia.Controls/Grid/DefinitionBase.cs]: @grokys This function is being called all over the place, basically i need both PrivateSharedSizeScope and SharedSizeGroup not null when this runs so i resorted in placing calls on this method on both Measure/Arrange which is not ideal...

File: src/Avalonia.Controls/Grid/Grid.cs [hunk 4]
```diff
@@ -0,0 +1,2373 @@
...
+            {
+                if (_columnDefinitions == null)
+                {
+                    ColumnDefinitions = new ColumnDefinitions();
+                }
+
+                return _columnDefinitions;
+            }
+            set
+            {
```
Comment (jmacato) [inline src/Avalonia.Controls/Grid/Grid.cs]: @grokys This and `RowDefinition`'s setter are horrible but i ran out of ideas on how to properly mirror those two on their array counterparts...

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 5]
```diff
@@ -180,5 +183,267 @@ public void Changing_Child_Column_Invalidates_Measure()
...
+            root.Child = outerScope;
+
+            root.Measure(new Size(50, 50));
+            root.Arrange(new Rect(new Point(), new Point(50, 50)));
+            Assert.Equal(0, outerGrid.ColumnDefinitions[0].ActualWidth);
+        }
+
+        [Fact]
+        public void Size_Is_Propagated_Between_Rows_And_Columns()
+        {
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]: @grokys This baffles me sir as i really have no idea why this doesn't work at all

**Comment by jmacato** [comment]
@AvaloniaUI/core please don't update this branch to master until this second phase is done.

**Review by grokys — APPROVED** [review]
Looks good to me, it's pretty hard to review as it's a code dump from WPF, but gets a 👍 from me!

**Comment by grokys** [comment]
Ah, seems there's a failing test?

**Pull request merged**